app/packages/looker/src/elements/common/util.test.ts

```diff
@@ -0,0 +1,525 @@
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EEBB00");
+  });
+});
+
+describe("Primitive fields bubbles get correct color", () => {
+  const sampleValue = "apple";
+  const edgeSetting1 = [];
+  const edgeSetting2 = [
+    {
+      path: "str_field",
+    },
+  ];
+  const edgeSetting3 = [
+    {
+      path: "str_field",
+      fieldColor: null,
+    },
+  ];
+  const edgeSetting4 = [
+    {
+      path: "str_field",
+      fieldColor: color1,
+      valueColors: null,
+    },
+  ];
+
+  it("Color by field mode, color can get resolved correctly in field mode with proper fallback", () => {
+    const sampleInput = {
+      coloring: coloringByField,
+      path: "str_field",
+      value: sampleValue,
+    };
+    const sampleSetting = [
+      {
+        path: "str_field",
+        fieldColor: color1,
+      },
+    ];
+
+    // Correct case:
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: sampleSetting,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe(color1);
+
+    // Edge cases: fallback to default color of the field
+    const defaultColor = getColor(
+      coloringByField.pool,
+      coloringByField.seed,
+      "str_field"
+    );
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting1,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe(defaultColor);
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting2,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe(defaultColor);
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting3,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe(defaultColor);
+  });
+
+  it("Color by value mode, edge cases fallback correctly", () => {
+    const sampleInput = {
+      coloring: coloringByValue,
+      path: "str_field",
+      value: sampleValue,
+    };
+
+    // Edge cases: fallback to default color of the field
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting1,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EE6600");
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting2,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EE0000");
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting2,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EE0000");
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting2,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EE0000");
+    expect(
+      getAssignedColor({
+        ...sampleInput,
+        customizeColorSetting: edgeSetting3,
+        isValidColor: isValidColorMock,
+      })
+    ).toBe("#EE0000");
```

Comment:
nit: I would create and array of test objects and iterate through it and expect() in a loop maybe to refactor some of the repeating code